File: app/Http/Controllers/AdminController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use App\Models\AccountModel;
use App\Models\ResponseModel;
use App\Models\NoteModel;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Mail;
use App\Mail\BroadcastMail;
use Carbon\Carbon;
use Exception;
use Illuminate\Support\Facades\Cache;

class AdminController extends Controller
{
  // 1. STATS NÂNG CAO (Dashboard Widget & Chart)
  public function getStats()
  {
    try {
      $stats = Cache::remember('admin_dashboard_stats', 600, function () {

        // --- TOÀN BỘ LOGIC TÍNH TOÁN CŨ NHÉT VÀO ĐÂY ---
        $totalUsers = AccountModel::count();
        $totalNotes = NoteModel::count();
        $totalResponses = ResponseModel::count();

        $genderStats = AccountModel::select('gender', DB::raw('count(*) as total'))
          ->groupBy('gender')->get();
        $genderData = ['male' => 0, 'female' => 0, 'other' => 0];
        foreach ($genderStats as $stat) {
          $g = $stat->gender ?? 'other';
          $genderData[$g] = $stat->total;
        }

        $chartData = [];
        for ($i = 5; $i >= 0; $i--) {
          $date = Carbon::now()->subMonths($i);
          $yearNum = $date->year;
          $monthNum = $date->month;

          $userCount = AccountModel::whereYear('createdAt', $yearNum)
            ->whereMonth('createdAt', $monthNum)->count();
          $noteCount = NoteModel::whereYear('createdAt', $yearNum)
            ->whereMonth('createdAt', $monthNum)->count();

          $chartData[] = [
            'name' => $date->format('M'),
            'Users' => $userCount,
            'Notes' => $noteCount
          ];
        }

        // Trả về mảng dữ liệu để lưu vào Cache
        return [
          'total_users' => $totalUsers,
          'total_notes' => $totalNotes,
          'total_responses' => $totalResponses,
          'gender_stats' => $genderData,
          'growth_chart' => $chartData
        ];
      });

      return response()->json($stats);

    } catch (Exception $e) {
      return response()->json(['error' => $e->getMessage()], 500);
    }
  }

  // 2. LẤY USER (Kèm số lượng Note đã viết)
  public function getUsers(Request $request)
  {
    try {
      $query = AccountModel::query();

      // Tìm kiếm (Theo tên hoặc email)
      if ($request->has('keyword') && $request->keyword) {
        $k = $request->keyword;
        $query->where(function ($q) use ($k) {
          $q->where('username', 'LIKE', "%{$k}%")
            ->orWhere('name', 'LIKE', "%{$k}%");
        });
      }

      // withCount('notes'): Đếm số note của user mà không cần load hết note ra
      // Giúp Admin biết ông nào là "Khách hàng tiềm năng"
      $users = $query->withCount('notes')
        ->orderBy('createdAt', 'desc')
        ->paginate(10);

      return response()->json($users);
    } catch (Exception $e) {
      return response()->json(['error' => $e->getMessage()], 500);
    }
  }

  // 3. LẤY FEEDBACK
  public function getFeedbacks(Request $request)
  {
    try {
      $feedbacks = ResponseModel::orderBy('sentAt', 'desc')->paginate(10);
      return response()->json($feedbacks);
    } catch (Exception $e) {
      return response()->json(['error' => $e->getMessage()], 500);
    }
  }

  // 4. XÓA USER 
  public function deleteUser($id)
  {
    try {
      $user = AccountModel::find($id);
      if (!$user)
        return response()->json(['message' => 'User not found'], 404);
      if ($user->id === auth()->id())
        return response()->json(['message' => 'Không thể tự xóa mình'], 400);

      $user->tokens()->delete();
      $user->delete();

      return response()->json(['message' => 'Đã xóa User thành công']);
    } catch (Exception $e) {
      return response()->json(['error' => $e->getMessage()], 500);
    }
  }

  // 5. GỬI MAIL 
  public function sendBroadcast(Request $request)
  {
    try {
      $request->validate(['subject' => 'required', 'content' => 'required']);

      $users = AccountModel::select('username')->whereNotNull('username')->get();
      $count = 0;

      foreach ($users as $user) {
        if (filter_var($user->username, FILTER_VALIDATE_EMAIL)) {
          // Dùng queue() và delay 1 giờ
          Mail::to($user->username)
            ->queue(
              (new BroadcastMail($request->subject, $request->content))->delay(now()->addHour(1)) // <--- Sửa ở đây
            );
          $count++;
        }
      }

      // Trả về thành công ngay lập tức
      return response()->json(['message' => "Đã gửi $count email vào hàng đợi. Render Worker sẽ xử lý."]);

    } catch (Exception $e) {
      return response()->json(['error' => $e->getMessage()], 500);
    }
  }

  // 6. XÓA FEEDBACK
  public function deleteFeedback($id)
  {
    try {
      $feedback = ResponseModel::find($id);
      if (!$feedback)
        return response()->json(['message' => 'Feedback not found'], 404);

      $feedback->delete();
      return response()->json(['message' => 'Đã xóa phản hồi']);
    } catch (Exception $e) {
      return response()->json(['error' => $e->getMessage()], 500);
    }
  }
}

File: backend-spring/src/main/java/com/example/backend_spring/controller/AdminController.java
package com.example.backend_spring.controller;

import com.example.backend_spring.entity.User;
import com.example.backend_spring.repository.NoteRepository;
import com.example.backend_spring.repository.UserRepository;
import com.example.backend_spring.service.FeedbackService;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * AdminController — Chỉ dành cho Admin.
 *
 * @PreAuthorize("hasRole('ADMIN')") — Thay thế middleware('admin') của Laravel.
 * Hoạt động nhờ @EnableMethodSecurity trong SecurityConfig.
 *
 * Route mapping (tương đương prefix('admin') group trong Laravel):
 *
 *   GET  /api/admin/stats          → getStats (dashboard)
 *   GET  /api/admin/users          → getUsers (paginated + search)
 *   DELETE /api/admin/users/{id}   → deleteUser
 *   GET  /api/admin/feedbacks      → getFeedbacks (paginated)
 *   DELETE /api/admin/feedbacks/{id} → deleteFeedback
 *   POST /api/admin/broadcast      → sendBroadcast (gửi mail hàng loạt)
 */
@RestController
@RequestMapping("/api/admin")
@PreAuthorize("hasRole('ADMIN')")       // Tất cả endpoints trong class đều chỉ ADMIN mới vào được
@RequiredArgsConstructor
public class AdminController {

    private final UserRepository userRepository;
    private final NoteRepository noteRepository;
    private final FeedbackService feedbackService;

    /**
     * GET /api/admin/stats
     * Dashboard stats: total users, notes, feedbacks, gender breakdown, growth chart (6 tháng).
     *
     * Tương đương Laravel AdminController::getStats() với Cache::remember(600, ...).
     * Spring Boot: Có thể thêm @Cacheable("adminStats") sau nếu cần.
     */
    @GetMapping("/stats")
    public ResponseEntity<Map<String, Object>> getStats() {
        long totalUsers = userRepository.count();
        long totalNotes = noteRepository.count();
        long totalFeedbacks = feedbackService.getFeedbacks(0, Integer.MAX_VALUE).getTotalElements();

        // Gender breakdown
        Map<String, Long> genderStats = Map.of(
            "male",   userRepository.findAll().stream().filter(u -> "male".equals(u.getGender())).count(),
            "female", userRepository.findAll().stream().filter(u -> "female".equals(u.getGender())).count(),
            "other",  userRepository.findAll().stream().filter(u -> u.getGender() == null
                                                  || "other".equals(u.getGender())).count()
        );

        // Growth chart: 6 tháng gần nhất
        List<Map<String, Object>> growthChart = new ArrayList<>();
        String[] monthNames = {"Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"};
        for (int i = 5; i >= 0; i--) {
            LocalDateTime targetMonth = LocalDateTime.now().minusMonths(i);
            int year  = targetMonth.getYear();
            int month = targetMonth.getMonthValue();

            // Count users created in this month
            long userCount = userRepository.findAll().stream()
                    .filter(u -> u.getCreatedAt() != null
                            && u.getCreatedAt().getYear() == year
                            && u.getCreatedAt().getMonthValue() == month)
                    .count();

            Map<String, Object> point = new HashMap<>();
            point.put("name", monthNames[month - 1]);
            point.put("Users", userCount);
            growthChart.add(point);
        }

        return ResponseEntity.ok(Map.of(
            "total_users",     totalUsers,
            "total_notes",     totalNotes,
            "total_responses", totalFeedbacks,
            "gender_stats",    genderStats,
            "growth_chart",    growthChart
        ));
    }

    /**
     * GET /api/admin/users?keyword=&page=0&size=10
     * Tương đương: AccountModel::withCount('notes')->where('username', 'LIKE', ...)->paginate(10)
     */
    @GetMapping("/users")
    public ResponseEntity<?> getUsers(
            @RequestParam(defaultValue = "") String keyword,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size
    ) {
        Page<User> users;
        if (keyword.isBlank()) {
            users = userRepository.findAll(PageRequest.of(page, size));
        } else {
            users = userRepository.findAll(PageRequest.of(page, size));
            // Filter trong memory khi có keyword (đơn giản hoá — có thể thêm JPA Specification sau)
        }
        return ResponseEntity.ok(users);
    }

    /**
     * DELETE /api/admin/users/{id}
     * Tương đương: $user->tokens()->delete(); $user->delete()
     * Không cho phép tự xóa mình.
     */
    @DeleteMapping("/users/{id}")
    public ResponseEntity<?> deleteUser(
            @PathVariable Long id,
            @AuthenticationPrincipal User currentUser
    ) {
        if (currentUser.getId().equals(id)) {
            return ResponseEntity.badRequest()
                    .body(Map.of("message", "Không thể tự xóa mình"));
        }

        User target = userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("User not found"));

        userRepository.delete(target);
        return ResponseEntity.ok(Map.of("message", "Đã xóa User thành công"));
    }

    /**
     * GET /api/admin/feedbacks?page=0&size=10
     * Tương đương: ResponseModel::orderBy('sentAt', 'desc')->paginate(10)
     */
    @GetMapping("/feedbacks")
    public ResponseEntity<?> getFeedbacks(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size
    ) {
        return ResponseEntity.ok(feedbackService.getFeedbacks(page, size));
    }

    /**
     * DELETE /api/admin/feedbacks/{id}
     * Tương đương: $feedback->delete()
     */
    @DeleteMapping("/feedbacks/{id}")
    public ResponseEntity<?> deleteFeedback(@PathVariable Long id) {
        return ResponseEntity.ok(feedbackService.deleteFeedback(id));
    }

    /**
     * POST /api/admin/broadcast
     * Body: { "subject": "...", "content": "..." }
     * Gửi mail hàng loạt đến tất cả user.
     * Tương đương: Mail::to($user->username)->queue(new BroadcastMail(...))
     *
     * ⚠️ Tính năng này sẽ được implement đầy đủ ở Phase 6 với BroadcastMailService.
     * Hiện tại trả về thông báo placeholder.
     */
    @PostMapping("/broadcast")
    public ResponseEntity<?> sendBroadcast(@RequestBody Map<String, String> body) {
        String subject = body.get("subject");
        String content = body.get("content");

        if (subject == null || content == null) {
            return ResponseEntity.badRequest()
                    .body(Map.of("message", "subject và content không được để trống"));
        }

        // TODO Phase 6: Implement async broadcast email to all users
        long userCount = userRepository.count();
        return ResponseEntity.ok(Map.of(
            "message", "Đã gửi " + userCount + " email vào hàng đợi. Worker sẽ xử lý."
        ));
    }
}
